designs/2018-processors-improvements/README.md

```diff
@@ -0,0 +1,241 @@
+
+```js
+module.exports = {
+    plugins: ["markdown"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    overrides: [
+        {
+            files: ["*.md"],
+            processor: "markdown/markdown"
+        }
+    ]
+};
+```
+
+In this configuration, the named processor `markdown` from the `eslint-plugin-markdown` plugin is applied only to files with the extension `*.md`. This mimics the behavior of the extension-named processor that `eslint-plugin-markdown` currently exports.
+
+The `CLIEngine` method `getConfigForFile()` would automatically get the correct value for `processor` when the configuration is calculated.
+
+**Note:** End-users would still need to provide `--ext .md` on the command line for ESLint to automatically lint files ending with `.md`. No part of this proposal changes the behavior of `--ext`.
+
+### Virtual Filenames for Code Blocks
+
+The `preprocess()` method inside of a processor can now return an array containing either strings (the current behavior) or an object with two properties:
+
+* `filename` - (string) a virtual file filename to map the code block to (cannot contain `\` or `/` characters)
+* `text` - (string) the source code text
+
+For example:
+
+```js
+module.exports = {
+    preprocess(text, filename) {
+        
+        const { jsText1, jsText2 } = doSomething(text, filename);
+        
+        return [
+            {
+                filename: "1.js",
+                text: jsText1
+            },
+            {
+                filename: "2.js",
+                text: jsText2
+            }
+        ]
+    }
+};
+```
+
+This processor returns two code blocks containing JavaScript code. Each code block is given a virtual filename ending with `.js`.
+
+When a `preprocess()` method returns an object with a `filename` property, `CLIEngine` will call `getConfigForFile()` on the `filename` property to determine the correct configuration for the code block, which includes whether another processor should be run on the code block (matched by file extension).
+
+When a `preprocess()` method returns only a string, `CLIEngine` will interpret that as a JavaScript file and call `getConfigForFile()` using the parent file's filename appended with `".js"` (in case the parent filename does not end with `".js"`) to determine the correction configuration for the code block using this algorithm:
+
+```js
+const codeBlocks = processor.preprocess(text, filename).map(item => {
+    if (typeof item === "string") {
+        return {
+            filename: filename + ".js",
+            text: item
+        };
+    } else {
+        return item;
+    }
+});
+```
+
+**Note:** Any virtual filename that does not have an extension already specified by `--ext` will be ignored and will not be linted or autofixed.
```

Comment:
Another concern that I thought of last night: right now, the `filename` parameter of `preprocess()` is guaranteed to be unique each time it is called. With the current proposal, it's possible for `filename` to not be unique because we allow `preprocess()` to assign whatever filename it wants. 

I think we need to do one of these:

1. Force `preprocess()` to only assign a file extension and not a full filename. We would then make each code block have a name of `parentFilename + index + ext`.
2. Use the returned filename from `preprocess()` as just part of the full filename, so `parentFilename + index + preprocessFilename`.

(We could also insert `block` or `codeblock` into each filename to make it easier to use with `overrides`.)

Reply:
@mysticatea thoughts on this?

Reply:
The `parentFilename + index + preprocessFilename` sounds good. We can mention the `preprocessFilename` can be only file extension in the document. If a code block has that name, `preprocess` can return the name.

For example:

- `{code: "...", filename: "js"}` → `parentFilename.md/0.js`
- `{code: "...", filename: "child.js"}` → `parentFilename.md/0.child.js`

If we use `path.sep` at between the parent filename and preprocess filename, it will not conflict with other existing files.

Reply:
I'd like to stick with just using a filename, as I think it's a bit confusing to allow either the extension or the filename in the same field.

I'll update the proposal for this part.

Reply:
I've updated the proposal, please take a look at this section.